endpoints/openrtb2/auction.go

```diff
@@ -336,111 +336,136 @@ func parseTimeout(requestJson []byte, defaultTimeout time.Duration) time.Duratio
 // Preference is given to parameters at imp[].ext level over req.ext level.
 // Parameters at req.ext level are propagated to adapters as is without any validation.
 func mergeBidderParams(req *openrtb_ext.RequestWrapper) error {
-	reqBidderParams, err := adapters.ExtractReqExtBidderParams(req.BidRequest)
+	reqBidderParams, err := adapters.ExtractReqExtBidderParamsEmbeddedMap(req.BidRequest)
 	if err != nil {
-		return err
+		return nil
+	}
+
+	if len(reqBidderParams) == 0 {
+		return nil
 	}
 
-	impCpy := make([]openrtb2.Imp, 0, len(req.BidRequest.Imp))
-	for _, imp := range req.BidRequest.Imp {
+	imps := make([]openrtb2.Imp, 0, len(req.BidRequest.Imp))
+	for impIndex, imp := range req.BidRequest.Imp {
 		updatedImp := imp
 
 		if len(imp.Ext) == 0 {
-			impCpy = append(impCpy, updatedImp)
+			imps = append(imps, updatedImp)
 			continue
 		}
 
-		var impExt map[string]map[string]json.RawMessage
-		err := json.Unmarshal(imp.Ext, &impExt)
-		if err != nil {
-			return err
+		var impExt map[string]json.RawMessage
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			// imp.ext should be parsable, but we haven't validated yet so there may be another issue. ignore it here
+			// and let the validation catch it later on.
+			imps = append(imps, updatedImp)
+			continue
 		}
 
 		//merges bidder parameters passed at req.ext level with imp[].ext level.
-		err = addMissingReqExtParamsInImpExt(impExt, reqBidderParams)
+		impExtModified, err := addMissingReqExtParamsInImpExt(impExt, reqBidderParams)
 		if err != nil {
-			return err
+			return fmt.Errorf("error processing bidder parameters for imp[%d]: %s", impIndex, err.Error())
 		}
 
 		//merges bidder parameters passed at req.ext level with imp[].ext.prebid.bidder level.
-		err = addMissingReqExtParamsInImpExtPrebid(impExt, reqBidderParams)
+		impExtPrebidModified, err := addMissingReqExtParamsInImpExtPrebid(impExt, reqBidderParams)
```

Comment:
Given that `request.imp.ext.prebid.bidder.BIDDER` takes precedence over `request.imp.ext.BIDDER`, does this become an "either-or" situation? If a given `imp.ext` comes with a `prebid` field, should we only call `addMissingReqExtParamsInImpExtPrebid` and not both `addMissingReqExtParamsInImpExtPrebid` and `addMissingReqExtParamsInImpExt`? If the `imp.ext`s that come with a `prebid` field will ignore `request.imp.ext.BIDDER` values anyways, maybe it'd be a desirable optimization to not copy them to `request.imp.ext.BIDDER` at all.
```
1124 func (deps *endpointDeps) validateImpExt(imp *openrtb2.Imp, aliases map[string]string, impIndex int) []error {
1125     errL := []error{}
            .
            .
1134
1135     // Prefer bidder params from request.imp.ext.prebid.bidder.BIDDER over request.imp.ext.BIDDER
1136     // to avoid confusion beteween prebid specific adapter config and other ext protocols.
1137     if extPrebidJSON, ok := bidderExts[openrtb_ext.PrebidExtKey]; ok {
1138         var extPrebid openrtb_ext.ExtImpPrebid
1139         if err := json.Unmarshal(extPrebidJSON, &extPrebid); err == nil && extPrebid.Bidder != nil {
1140             for bidder, ext := range extPrebid.Bidder {
1141                 if ext == nil {
1142                     continue
1143                 }
1144                 bidderExts[bidder] = ext
1145             }
1146         }
1147     }
1148
1149     /* Process all the bidder exts in the request */
endpoints/openrtb2/auction.go
```
If implemented, the change would make `mergeBidderParams` behave differently, and a test case like the one shown below would expect a different outcome:
```
1149   func TestMergeBidderParams(t *testing.T) {
1150       testCases := []struct {
1151           description         string
1152           givenRequest        openrtb2.BidRequest
1153           expectedRequestImps []openrtb2.Imp
1154       }{
1155   *- 55 lines: {-------------------------------------------------------------------------------
1210           {
1211 -             description: "One Imp - imp.ext + imp.ext.prebid.bidder Modified - Same Bidder",
     +             description: "One Imp - modify imp.ext.prebid.bidder ONLY - Same Bidder",
1212               givenRequest: openrtb2.BidRequest{
1213                   Imp: []openrtb2.Imp{{ID: "1", Ext: json.RawMessage(`{"bidder1":{"a":1},"prebid":{"bidder":{"bidder1":{"a":"one"}}}}`)}},
1214                   Ext: json.RawMessage(`{"prebid":{"bidderparams":{"bidder1":{"b":2}}}}`),
1215               },
1216 -             expectedRequestImps: []openrtb2.Imp{{ID: "1", Ext: json.RawMessage(`{"bidder1":{"a":1,"b":2},"prebid":{"bidder":{"bidder1":{"a":"one","b":2}}}}`)}},
     +             expectedRequestImps: []openrtb2.Imp{{ID: "1", Ext: json.RawMessage(`{"bidder1":{"a":1},"prebid":{"bidder":{"bidder1":{"a":"one","b":2}}}}`)}},
1217           },
endpoints/openrtb2/auction_test.go
```
Unless I'm missing something and those values are still needed in `request.imp.ext.BIDDER`. Thoughts?

Reply:
> Given that request.imp.ext.prebid.bidder.BIDDER takes precedence over request.imp.ext.BIDDER, does this become an "either-or" situation?

Yes. Later code will pick one or the other, but not both. Referencing the precedence logic here would increase it's complexity for IMHO minimal processing savings.

I'm not sure why it was designed this way, but separating the merging pf params from the decision on which params to use seems like a good isolation of responsibilities.

This is pre-existing behavior. The reason I didn't refactor that in this PR is that it's not required for `imp.ext.gpid` to function, which is the goal of this PR.

Reply:
makes sense. It can be refactored separately